catalog/transformer/catalog-transformer-streaming-lib/src/main/java/org/codice/ddf/transformer/xml/streaming/lib/MetacardTypeRegister.java

```diff
@@ -0,0 +1,142 @@
+
+    private Map<String, ServiceRegistration> metacardTypeServiceRegistrations = new HashMap<>();
+
+    private List<SaxEventHandlerFactory> saxEventHandlerFactories = new ArrayList<>();
+
+    private DynamicMetacardType metacardType;
+
+    private String id = "DEFAULT_ID";
+
+    public synchronized void bind(SaxEventHandlerFactory saxEventHandlerFactory) {
+        saxEventHandlerFactories.add(saxEventHandlerFactory);
+        setMetacardType();
+    }
+
+    public synchronized void unbind(SaxEventHandlerFactory saxEventHandlerFactory) {
+        saxEventHandlerFactories.remove(saxEventHandlerFactory);
+        setMetacardType();
+    }
+
+    /**
+     * Defines a {@link DynamicMetacardType} based on component Sax Event Handler Factories
+     * and what attributes they populate
+     *
+     * @return a DynamicMetacardType that describes the type of metacard that is created in this transformer
+     */
+    private synchronized void setMetacardType() {
+        Set<AttributeDescriptor> attributeDescriptors = new HashSet<>();
+
+        if (saxEventHandlerFactories != null) {
+            for (SaxEventHandlerFactory factory : saxEventHandlerFactories) {
+                attributeDescriptors.addAll(factory.getSupportedAttributeDescriptors());
+            }
+        }
+        attributeDescriptors.addAll(BasicTypes.BASIC_METACARD.getAttributeDescriptors());
+
+        DynamicMetacardType dynamicMetacardType = new DynamicMetacardType(attributeDescriptors, id);
+        registerMetacardType(dynamicMetacardType);
+        metacardType = dynamicMetacardType;
+    }
+
+    public MetacardType getMetacardType() {
+        if (metacardType == null) {
+            setMetacardType();
+        }
+
+        return metacardType;
+    }
+
+    private synchronized void registerMetacardType(MetacardType metacardType) {
+        unregisterMetacardType(metacardType);
+
+        Dictionary serviceProperties = new Hashtable();
+        serviceProperties.put("name", metacardType.getName());
+
+        ServiceRegistration serviceRegistration =
+                getContext().registerService(MetacardType.class.getName(),
+                        metacardType,
+                        serviceProperties);
+
+        this.metacardTypeServiceRegistrations.put(metacardType.getName(), serviceRegistration);
+    }
+
+    private synchronized void unregisterMetacardType(MetacardType metacardType) {
+        if (metacardType == null) {
+            return;
+        }
+
+        ServiceRegistration serviceRegistration =
+                metacardTypeServiceRegistrations.get(metacardType.getName());
+        if (serviceRegistration != null) {
+            serviceRegistration.unregister();
+            metacardTypeServiceRegistrations.remove(metacardType.getName());
+        }
+    }
+
+    public synchronized void setId(String id) {
+        this.id = id;
+    }
+}
+
+class DynamicMetacardType implements MetacardType {
```

Comment:
I could try changing it to that. The previous implementation just used this private inner class so I just moved it around